Add new workshop to replace old sign and asphalt workshop.

Thomas--S opened the pull request:
Better usability and cheaper crafting.

The old workshops consumed way too much dye in my opinion. Working with them in a survival world was really no fun.

@webD97 and @cheapie please review this. I would be very glad if this could be merged.

I submit this as a PR, as this is quite a big change.

Thanks!

Thomas--S commented on streets_workshop/init.lua:
```diff
@@ -0,0 +1,648 @@
+				and inv:get_stack("sign_input", 1):get_count() >= 1 then
+			return true
+		end
+		minetest.chat_send_all("test")
```
I'll remove this forgotten debug output later.

webD97 commented:
@Thomas--S Question: What do the dye indicators show? The level relative to the complete dye stack in it or the level of a single dye item? It either seems to jump around or I don't get the logic behind it...

webD97 commented:
Do we need rotated versions of every single item in it? This basically explodes the number of nodes registered to the engine. As you might have seen on the forums, people start complaining about long loading times.

webD97 commented:
@Thomas--S I did my best to test everything and it seems to work perfectly. I really like the user interface. One addition I'd like to see: When selecting a new category for crafting, let's say from "center lines" to "EU regulatory", it should automatically clear the selection. Otherwise it will continue crafting while the user actually does not want it to continue crafting.

Thomas--S commented:
1) The dye indicators show the level of the internal dye storage. The internal dye storage can be filled with up to 4 dyes. These get removed from the stack.

2) The rotated items aren't added in this PR, they already existed before. Additionally, we don't have a rotated version for every single item, but only for those were rotation is needed. The rotated items are useful when using the moreblocks version of the asphalt nodes.

3) Regarding the lag: In the future, I plan to migrate the roadmarkings to meta-set nodedef. In June, I already started a PR for meta-set nodedef, but I fear I won't have the time to maintain and complete it.
I already have a little bit different concept in mind, maybe I'll program this when I have time again.

4) Regarding the automatic clearing of the selection when changing section: I personally prefer the current behaviour, but I can understand that other people prefer your version. As a compromise, I suggest to add a new setting in the setup page. Your suggestion will be the default value.

5) I could add a trashcan slot in a new tab, if this is ok for you.

Thank you very much for your review! I suggest not to merge this before the beginning of next week, as I won't be able to change these things earlier.

Thomas--S commented:
> One addition I'd like to see: When selecting a new category for crafting, let's say from "center lines" to "EU regulatory", it should automatically clear the selection. Otherwise it will continue crafting while the user actually does not want it to continue crafting.

One question:
Are you talking about changing category (roadmarkings, signs, asphalt) or about changing section (EU regulatory, EU mandatory, Center lines, Side lines, ...)

webD97 commented:
Okay, thank you.

> The dye indicators show the level of the internal dye storage. The internal dye storage can be filled with up to 4 dyes. These get removed from the stack.

Maybe I'm still getting the logic wrong. While trying out the crafting of multiple items, the indicator goes down to somehwere between 80 and 90% and then back up to 100%. I don't know how to describe it. If you can assure that the behaviour is correct, that's fine for me. Maybe I'm just a bit slow today ;)

> The rotated items aren't added in this PR, they already existed before. Additionally, we don't have a rotated version for every single item, but only for those were rotation is needed. The rotated items are useful when using the moreblocks version of the asphalt nodes.

Correct. Maybe this is the wrong place to discuss this. It has just caught my eye again ;)

> Regarding the lag: In the future, I plan to migrate the roadmarkings to meta-set nodedef. In June, I already started a PR for meta-set nodedef, but I fear I won't have the time to maintain and complete it.
I already have a little bit different concept in mind, maybe I'll program this when I have time again.

Very interesting. I'd really like to see the concept, but please take all the time you need.

> Regarding the automatic clearing of the selection when changing section: I personally prefer the current behaviour, but I can understand that other people prefer your version. As a compromise, I suggest to add a new setting in the setup page. Your suggestion will be the default value.

Cool, thanks again.

> I could add a trashcan slot in a new tab, if this is ok for you.

I don't think a new tab would be sufficient. Better than none at all, but still not enough. The idea behind it is that it should be accessible as fast as possible, so it should be directly on the crafting tab.

> One question: Are you talking about changing category (roadmarkings, signs, asphalt) or about changing section (EU regulatory, EU mandatory, Center lines, Side lines, ...)?

Changing section.

Thomas--S commented:
> Maybe I'm still getting the logic wrong. While trying out the crafting of multiple items, the indicator goes down to somehwere between 80 and 90% and then back up to 100%. I don't know how to describe it. If you can assure that the behaviour is correct, that's fine for me. Maybe I'm just a bit slow today ;)

You can think of the indicator as kind of a "box" where dye is stored. One dye item fills this box 25%. As soon as there is place for the next dye (i.e. when the storage goes below 75%) the next dye of the slot is used to fill the storage "box". I hope that this description is understandable; this concept may be a bit hard to explain. If needed, I could maybe draw a little picture showing the concept.

To understand those storage/indicator boxes better, maybe it would be a good idea not to place a whole stack in the input slot, but only a single dye at a time.

---

> I don't think a new tab would be sufficient. Better than none at all, but still not enough. The idea behind it is that it should be accessible as fast as possible, so it should be directly on the crafting tab.

I'll try to find a good place for this. The problem is, that the space in the crafting tab is very limited, especially for roadmarkings and signs.

---

_Regarding the meta-set nodedef:_
You can find the current approach under minetest/minetest#5919. I think that this is a bit too complicated and hard to maintain, so I thought about the following:
Add a new `get_texture(pos)` function to the nodedef. By default, this function returns the `tiles` field. This function could then be programmed to use the meta to decide which texture to use. The problem is, that this is currently only an idea; I didn't look at the source code until now. I hope I will find a way sometime to implement this.

Thanks for your answer and your interest. 👍

webD97 commented:
Thanks, now I understand it ^^

Thomas--S commented:
**PR updated:**

* Removed forgotten debug output
* Added trash can slot in craft tab
* Setting for clearing selection on section change

I think this is ready for merge now.

webD97 commented:
Good work, thank you! I'd say we'll wait two more days to see if @cheapie will review it as well. Otherwise we'll merge on September 1.

cheapie commented:
I've kind of stopped developing this mod and wasn't planning on reviewing this. You two have fun :P

webD97 commented:
Okay, thanks for the info

Thomas--S commented:
Thanks for merging!
